cs3api4lab/auth/check_auth_interceptor.py

```diff
@@ -0,0 +1,39 @@
+import http
+
+import cs3
+import grpc
+import cs3.rpc.code_pb2 as cs3code
+
+
+class CheckAuthInterceptor(grpc.UnaryUnaryClientInterceptor,
+                           grpc.UnaryStreamClientInterceptor,
+                           grpc.StreamUnaryClientInterceptor,
+                           grpc.StreamStreamClientInterceptor):
+    unauth_codes = {cs3code.CODE_UNAUTHENTICATED}
+
+    def __init__(self, log, authenticator):
+        self.log = log
+        self.authenticator = authenticator
+
+    def intercept_unary_unary(self, continuation, client_call_details, request):
+        return self._intercept_call(continuation, client_call_details, request)
+
+    def intercept_stream_unary(self, continuation, client_call_details, request_iterator):
+        return self._intercept_call(continuation, client_call_details, request_iterator)
+
+    def intercept_unary_stream(self, continuation, client_call_details, request):
+        return self._intercept_call(continuation, client_call_details, request)
+
+    def intercept_stream_stream(self, continuation, client_call_details, request_iterator):
+        return self._intercept_call(continuation, client_call_details, request_iterator)
+
+    def _intercept_call(self, continuation, client_call_details, request_or_iterator):
+        response = continuation(client_call_details, request_or_iterator)
+        self._check_result(response.code(), response.result())
+        return response
+
+    def _check_result(self, code, result):
+        if result is not None and result.status is not None and result.status.code in self.unauth_codes:
+            if self.log is not None:
+                self.log.info("Refresh auth token")
+            self.authenticator.refresh_token()
```

Comment:
I still don't understand why we intercept things if we don't re-issue the command.  But then you call refresh without the `client_id`.. So, in practice, this doesn't do anything?

Reply:
I raise 401 error when Reva return unauth code instead call method to refreshing token

Reply:
I don't understand your message. Can you explain?

Reply:
Now, if we checked tokens expire before making a request and refresh if is necessary. In this case, it very rare to have a response from Reva with an authentication error. If so, it is better to return 401 than try to refresh the token and add request repetition logic.

Reply:
But we do check them, no? When you ask for the token, it's validated before being returned, so we don't need this.

```python
    def authenticate(self):
        """
        The basic authenticate method, return IOP token or refresh if is not present or invalid.
        """
        if self.token is None or self._check_token(self.token) is False:
            self.refresh_token()

        return self.token
```

Reply:
So yes, return a 401 if it failed and that's it.

Reply:
In general "yes". It already depends on the implementation of "refresh_token" in the children class. For example, RevaPassword use values from config to log in at Reva, in failure raise 401 error; oAuth has a different way: check token, is not valid raise 401 error, login at Reva, in failure raise 401 error. If classes don't override "refresh_token" method immediately raise 401 error.